File: kotlin/meetup/src/main/kotlin/Meetup.kt
import java.time.DayOfWeek
import java.time.LocalDate

class Meetup(val month: Int, val year: Int) {
    fun day(dayOfWeek: DayOfWeek, schedule: MeetupSchedule): LocalDate {
        val dates = datesOfDay(dayOfWeek)
        return when (schedule) {
            MeetupSchedule.TEENTH -> dates.first { it.getDayOfMonth() in 13..19 }
            MeetupSchedule.LAST -> dates.toList().reversed().first()
            else -> dates.elementAt(schedule.offset - 1)
        }
    }

    private fun datesOfDay(dayOfWeek: DayOfWeek): Sequence<LocalDate>
        = generateSequence(firstDayOfMonth(dayOfWeek)) {
            val next = it.plusDays(7)
            if (next.getMonthValue() == month) next else null
        }

    private fun firstDayOfMonth(dayOfWeek: DayOfWeek): LocalDate {
        var date = LocalDate.of(year, month, 1)
        while (date.getDayOfWeek() != dayOfWeek) {
            date = date.plusDays(1)
        }
        return date
    }
}

File: ruby/meetup/_iterations/2/meetup.rb
require 'date'

class Meetup
  NTHS = { :first => 0, :second => 1, :third => 2, :fourth => 3, :last => -1 }

  attr_reader :month, :year

  def initialize(month, year)
    @month = month
    @year = year
  end

  def day(weekday, nth)
    weekdays = (Date.new(year, month, 1)..Date.new(year, month, -1)).filter(&:"#{weekday}?")
    return weekdays[NTHS[nth]] unless nth == :teenth
    weekdays.find { |d| (13..19) === d.day }
  end
end
